File: fix_pairs.rb
#!/usr/bin/env ruby

# Copyright 2016 Ryan Moore
# Contact: moorer@udel.edu
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

require "set"
require "trollop"
require "abort_if"

module CoreExtensions
  module Time
    def date_and_time fmt="%F %T.%L"
      Object::Time.now.strftime fmt
    end

    def time_it title="", logger=AbortIf.logger, run: true
      if run
        t = Object::Time.now

        yield

        time = Object::Time.now - t

        if title == ""
          msg = "Finished in #{time} seconds"
        else
          msg = "#{title} finished in #{time} seconds"
        end

        if logger
          logger.info msg
        else
          $stderr.puts msg
        end
      end
    end
  end
end

def index_fastq fname
  index = {}
  count = 0
  seq_num = 0
  seq_offset = 0
  qual_offset = 0
  name_id = ""
  header_offset = 0 # with the @

  File.open(fname, "rt") do |f|
    f.each_line do |line|
      STDERR.printf("Reading -- %d\r", seq_num) if (seq_num % 10000).zero?

      case count
      when 0 # header
        name_id = line.chomp.split(" ")[0][1..-1]
        seq_offset = f.tell
      when 1 # seq

      when 2 # desc
        qual_offset = f.tell
      when 3 # qual
        count = -1
        index[name_id] = [header_offset, seq_offset, qual_offset]
        # outf.puts [seq_num, seq_offset, qual_offset].join "\t"
        seq_num += 1
        header_offset = f.tell
      end

      count += 1
    end
  end

  AbortIf.logger.debug { "Num seqs in '#{fname}': #{index.count}" }

  index
end

def read_record f, idx_ary
  header_offset = idx_ary[0]
  seq_offset = idx_ary[1]
  qual_offset = idx_ary[2]

  f.seek header_offset, IO::SEEK_SET
  header = f.readline.chomp

  f.seek seq_offset, IO::SEEK_SET
  seq = f.readline.chomp

  f.seek qual_offset, IO::SEEK_SET
  qual = f.readline.chomp

  "#{header}\n#{seq}\n+\n#{qual}"
end

opts = Trollop.options do
  version "Version: fix_pairs.rb v0.1.0"
  banner <<-EOS

  Assumes that the paired reads have headers that match up until the
  first space like these:

    @SN741:746:HKFKLBCXX:1:1106:19267:2152 1:N:0:TGCGTAAC
    @SN741:746:HKFKLBCXX:1:1106:19267:2152 2:N:0:TGCGTAAC
    ^ from here...they match...till here ^

  This program does not check that every sequence has the proper
  Illumina forward or reverse tag. It assumes that you have the
  correct reads in the correct command line args.

  An index is used rather than reading seqs into memory.

  Usage:
    ruby fix_pairs.rb foward_seqs.1.fq reverse_seqs.2.fq outfile_base_name

  Output files (given the above command):
    outfile_base_name.1.fq <= surviving forward reads
    outfile_base_name.2.fq <= surviving reverse reads
    outfile_base_name.U.fq <= unpaired reads

  Options:
  EOS
end

include AbortIf
Time.extend CoreExtensions::Time

forward = ARGV[0]
reverse = ARGV[1]
outbase = ARGV[2]
for_outf = "#{outbase}.1.fq"
rev_outf = "#{outbase}.2.fq"
un_outf  = "#{outbase}.U.fq"
for_idx = nil
rev_idx = nil
for_keys = nil
rev_keys = nil
paired_keys = nil
for_unpaired_keys = nil
rev_unpaired_keys = nil

Time.time_it "Indexing fastQ files" do
  for_idx = index_fastq forward
  rev_idx = index_fastq reverse
end

Time.time_it "Finding paired IDs" do
  for_keys = Set.new for_idx.keys
  rev_keys = Set.new rev_idx.keys

  paired_keys = for_keys.intersection rev_keys


  for_unpaired_keys = for_keys - paired_keys
  rev_unpaired_keys = rev_keys - paired_keys

  AbortIf.logger.debug { "Num paired keys: #{paired_keys.count}" }
  AbortIf.logger.debug { "Num for only keys: #{for_unpaired_keys.count}" }
  AbortIf.logger.debug { "Num rev only keys: #{rev_unpaired_keys.count}" }
end

begin
  forf = File.open(forward, "rt")
  revf = File.open(reverse, "rt")

  Time.time_it "Writing paired reads" do
    File.open(for_outf, "w") do |foutf|
      File.open(rev_outf, "w") do |routf|
        n = 0
        paired_keys.each do |key|
          n+=1; STDERR.printf("Reading -- %d\r", n) if (n % 10000).zero?

          for_idx_ary = for_idx[key]
          rev_idx_ary = rev_idx[key]

          foutf.puts read_record forf, for_idx_ary
          routf.puts read_record revf, rev_idx_ary
        end
      end
    end
  end

  Time.time_it "Writing un-paired reads" do
    File.open(un_outf, "w") do |f|
      n = 0
      for_unpaired_keys.each do |key|
        n+=1; STDERR.printf("Reading -- %d\r", n) if (n % 10000).zero?

        for_idx_ary = for_idx[key]

        f.puts read_record forf, for_idx_ary
      end

      n = 0
      rev_unpaired_keys.each do |key|
        n+=1; STDERR.printf("Reading -- %d\r", n) if (n % 10000).zero?

        rev_idx_ary = rev_idx[key]

        f.puts read_record revf, rev_idx_ary
      end
    end
  end
ensure
  forf.close
  revf.close
end

AbortIf.logger.info { "Surviving forward seqs: '#{for_outf}'" }
AbortIf.logger.info { "Surviving reverse seqs: '#{for_outf}'" }
AbortIf.logger.info { "Unpaired seqs: '#{un_outf}'" }

File: fix_pairs.cc
// Copyright 2016 Ryan Moore
// Contact: moorer@udel.edu

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see
// <http://www.gnu.org/licenses/>.

// Compile with g++ -g -Wall --std=c++11 test.cc -o test

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

// see http://stackoverflow.com/questions/236129/split-a-string-in-c
std::string get_id(const std::string &header) {
  std::stringstream ss;
  ss.str(header);
  std::string id;

  std::getline(ss, id, ' ');
  return id;
}

char * check_asprintf(char * basename, const char * suffix)
{
  char * str;
  if (asprintf(&str, "%s%s", basename, suffix) == -1) {
    std::cerr << "ERROR -- Memory error in allocating file name for "
              << basename
              << suffix
              << std::endl;

    exit(2);
  }

  return str;
}

void check_open(std::ofstream * stream, char * name)
{
  stream->open(name);

  if (!stream->is_open()) {
    std::cerr << "Couldn't open "
              << name
              << " for writing"
              << std::endl;

    exit(3);
  }
}

int main(int argc, char *argv[])
{

  if (argc != 4) {
    std::cerr << "USAGE: "
              << argv[0]
              << " reads.1.fq reads.2.fq outfile_basename"
              << std::endl;

    return 1;
  }

  int lineno = -1;

  long int nf = 0;
  long int nr = 0;
  long int num_surviving_pairs = 0;
  long int num_broken_forward = 0;
  long int num_broken_reverse = 0;

  std::string line, header, seq, comment, qual, rec_str;

  std::ifstream forward (argv[1]);
  std::ifstream reverse (argv[2]);

  std::ofstream for_outf;
  std::ofstream rev_outf;
  std::ofstream unp_outf;

  std::unordered_map<std::string, std::string> recs;
  std::unordered_map<std::string, std::string>::const_iterator iter;

  char * for_outf_name;
  char * rev_outf_name;
  char * unp_outf_name;

  for_outf_name = check_asprintf(argv[3], ".1.fq");
  rev_outf_name = check_asprintf(argv[3], ".2.fq");
  unp_outf_name = check_asprintf(argv[3], ".U.fq");

  check_open(&for_outf, for_outf_name);
  free(for_outf_name);

  check_open(&rev_outf, rev_outf_name);
  free(rev_outf_name);

  check_open(&unp_outf, unp_outf_name);
  free(unp_outf_name);

  if (forward.is_open()) {
    while (std::getline(forward, line)) {
      ++lineno;

      if (lineno == 0) { // header
        header = line;
      } else if (lineno == 1) { // seq
        seq = line;
      } else if (lineno == 2) { // comment
        comment = line;
      } else if (lineno == 3) { // qual
        if (++nf % 10000 == 0) {
          std::cerr << "Reading forward -- "
                    << nf
                    << "\r";
        }

        qual = line;

        rec_str = header + "\n" + seq + "\n" + comment + "\n" + qual;

        // do the stuff
        recs.emplace(get_id(header), rec_str);

        // reset vals for next record
        lineno  = -1;
        header  = "";
        seq     = "";
        comment = "";
        qual    = "";
      }
    }
  } else {
    std::cerr << "ERROR -- cannot read "
              << argv[1]
              << std::endl;
    exit(4);
  }

  if (reverse.is_open()) {
    // do stuff
    while (std::getline(reverse, line)) {
      ++lineno;

      if (lineno == 0) { // header
        header = line;
      } else if (lineno == 1) { // seq
        seq = line;
      } else if (lineno == 2) { // comment
        comment = line;
      } else if (lineno == 3) { // qual
        if (++nr % 10000 == 0) {
          std::cerr << "Reading reverse -- "
                    << nr
                    << "\r";
        }

        qual = line;

        rec_str = header + "\n" + seq + "\n" + comment + "\n" + qual;

        // check for key
        iter = recs.find(get_id(header));

        if (iter != recs.end()) { // key found
          ++num_surviving_pairs;

          // print the record in recs in the forward file
          for_outf << iter->second << std::endl;

          // print this record in the reverse file
          rev_outf << rec_str << std::endl;

          // delete the id from the map cos it has already been printed
          recs.erase(iter);
        } else {
          ++num_broken_reverse;
          // print this record in the unpaired file
          unp_outf << rec_str << std::endl;
        }

        // reset vals for next record
        lineno  = -1;
        header  = "";
        seq     = "";
        comment = "";
        qual    = "";
      }
    }
  } else {
    std::cerr << "ERROR -- cannot read "
              << argv[2]
              << std::endl;

    exit(5);
  }

  // print the unpaired forward ids
  for (iter = recs.begin(); iter != recs.end(); ++iter) {
    ++num_broken_forward;

    unp_outf << iter->second
             << std::endl;
  }


  for_outf.close();
  rev_outf.close();
  unp_outf.close();

  std::cerr << "\n\n----Results-------------------------------------------------\n";
  std::cerr << "Num input forward reads:   " << nf << std::endl;
  std::cerr << "Num input reverse reads:   " << nr << std::endl;
  std::cerr << "Total input reads:         " << nr + nf << std::endl;
  std::cerr << std::endl;
  std::cerr << "Num surviving read pairs:  " << num_surviving_pairs << std::endl;
  std::cerr << "Num broken forward reads:  " << num_broken_forward << std::endl;
  std::cerr << "Num broken reverse reads:  " << num_broken_reverse << std::endl;
  std::cerr << "Total reads accounted for: " <<
    num_broken_reverse + num_broken_forward + (2 * num_surviving_pairs) << std::endl;
  std::cerr << "------------------------------------------------------------" << "\n\n\n";


  return 0;
}
